lib/umwelt/parser.ex

```diff
@@ -0,0 +1,46 @@
+defmodule Umwelt.Parser do
+  @moduledoc "Extracts metadata from AST"
+
+  alias Umwelt.Parser
+
+  def read_ast({:error, msg}), do: {:error, msg}
+
+  def read_ast({:ok, code}),
+    do: Code.string_to_quoted(code)
+
+  def parse({:ok, ast}),
+    do: parse(ast, []) |> index()
+
+  def parse({_, _, _} = ast, aliases),
+    do: Parser.Triple.parse(ast, aliases)
+
+  def parse({_, _} = ast, aliases),
+    do: Parser.Tuple.parse(ast, aliases)
+
+  def parse(ast, aliases) when is_list(ast),
+    do: ast |> Enum.map(&parse(&1, aliases))
+
+  def parse(ast, _aliases),
+    do: Parser.Literal.parse(ast)
+
+  defp index(modules) do
+    [root_module(modules) | inner_modules(modules)]
+    |> Enum.reduce(%{}, fn item, result ->
+      Map.put(result, context(item), item)
+    end)
```

Comment:
```suggestion
    Enum.reduce(
      [root_module(modules) | inner_modules(modules)],
      %{},
      &Map.put(&2, context(&1), &1)
    )
```

`Map.update/4` is always better in such a case, you don’t want to silently discard duplicates.